### PR TITLE
Update packages to use `satisfies` instead of `if "+variant" in spec:`

### PR DESCRIPTION
**Pull request by aurianer**
Related to [this comment](https://github.com/spack/spack/pull/44774#discussion_r1649148676), I just did a short script, not sure it covers all cases but it did quite some replacements :)

### CODE REVIEW

File: var/spack/repos/builtin/packages/conduit/package.py [hunk 1]
```diff
@@ -538,9 +538,9 @@ def hostconfig(self):
 
         cfg.write("# hdf5 from spack \n")
 
-        if "+hdf5" in spec:
+        if spec.satisfies("+hdf5"):
             cfg.write(cmake_cache_entry("HDF5_DIR", spec["hdf5"].prefix))
-            if "zlib-api" in spec:
+            if spec.satisfies("zlib-api"):
```
Comment (wdconinc) [inline var/spack/repos/builtin/packages/conduit/package.py]: This seems incorrect.
```suggestion
            if spec.satisfies("^zlib-api"):
```

File: var/spack/repos/builtin/packages/converge/package.py [hunk 2]
```diff
@@ -146,29 +146,29 @@ def install(self, spec, prefix):
         # The CONVERGE tarball comes with binaries for several MPI libraries.
         # Only install the binary that matches the MPI we are building with.
         with working_dir("l_x86_64/bin"):
-            if "~mpi" in spec:
+            if spec.satisfies("~mpi"):
                 converge = glob.glob("converge-*-serial*")
                 post_convert = glob.glob("post_convert_serial*")
-            elif "hp-mpi" in spec:
+            elif spec.satisfies("hp-mpi"):
```
Comment (wdconinc) [inline var/spack/repos/builtin/packages/converge/package.py]:
```suggestion
            elif spec.satisfies("^hp-mpi"):
```

File: var/spack/repos/builtin/packages/converge/package.py [hunk 3]
```diff
@@ -146,29 +146,29 @@ def install(self, spec, prefix):
...
                 converge = glob.glob("converge-*-hpmpi*")
                 # No HP-MPI version of post_convert
                 post_convert = glob.glob("post_convert_serial*")
             elif "intel-mpi" in spec or "intel-parallel-studio+mpi" in spec:
                 converge = glob.glob("converge-*-intel*")
                 # No Intel MPI version of post_convert
                 post_convert = glob.glob("post_convert_serial*")
-            elif "mpich" in spec:
+            elif spec.satisfies("mpich"):
```
Comment (wdconinc) [inline var/spack/repos/builtin/packages/converge/package.py]:
```suggestion
            elif spec.satisfies("^mpich"):
```

File: var/spack/repos/builtin/packages/converge/package.py [hunk 4]
```diff
@@ -146,29 +146,29 @@ def install(self, spec, prefix):
...
                 converge = glob.glob("converge-*-mpich*")
                 post_convert = glob.glob("post_convert_mpich*")
-            elif "mvapich2" in spec:
+            elif spec.satisfies("mvapich2"):
```
Comment (wdconinc) [inline var/spack/repos/builtin/packages/converge/package.py]:
```suggestion
            elif spec.satisfies("^mvapich2"):
```

File: var/spack/repos/builtin/packages/converge/package.py [hunk 5]
```diff
@@ -146,29 +146,29 @@ def install(self, spec, prefix):
...
                 converge = glob.glob("converge-*-mvapich*")
                 # MVAPICH2 hasn't been supported since CONVERGE
                 # came with a single serial post_convert
                 post_convert = glob.glob("post_convert")
-            elif "openmpi" in spec:
+            elif spec.satisfies("openmpi"):
```
Comment (wdconinc) [inline var/spack/repos/builtin/packages/converge/package.py]:
```suggestion
            elif spec.satisfies("^openmpi"):
```

File: var/spack/repos/builtin/packages/converge/package.py [hunk 6]
```diff
@@ -146,29 +146,29 @@ def install(self, spec, prefix):
...
                 converge = glob.glob("converge-*-o*mpi*")
                 post_convert = glob.glob("post_convert_o*mpi*")
-            elif "platform-mpi" in spec:
+            elif spec.satisfies("platform-mpi"):
```
Comment (wdconinc) [inline var/spack/repos/builtin/packages/converge/package.py]:
```suggestion
            elif spec.satisfies("^platform-mpi"):
```
but this does not seem to be a package...

File: var/spack/repos/builtin/packages/cp2k/package.py [hunk 7]
```diff
@@ -453,31 +453,31 @@ def edit(self, pkg, spec, prefix):
...
 
-        if "@7:" in spec:  # recent versions of CP2K use C++14 CUDA code
+        if spec.satisfies("@7:"):  # recent versions of CP2K use C++14 CUDA code
             cxxflags.append(pkg.compiler.cxx14_flag)
             nvflags.append(pkg.compiler.cxx14_flag)
 
         ldflags.append(fftw.libs.search_flags)
 
-        if "superlu-dist@4.3" in spec:
+        if spec.satisfies("superlu-dist@4.3"):
```
Comment (wdconinc) [inline var/spack/repos/builtin/packages/cp2k/package.py]:
```suggestion
        if spec.satisfies("^superlu-dist@4.3"):
```

File: var/spack/repos/builtin/packages/cp2k/package.py [hunk 8]
```diff
@@ -651,10 +651,10 @@ def edit(self, pkg, spec, prefix):
             libs.extend(mpi)
             libs.extend(pkg.compiler.stdcxx_libs)
 
-            if "+mpi_f08" in spec:
+            if spec.satisfies("+mpi_f08"):
                 cppflags.append("-D__MPI_F08")
 
-            if "wannier90" in spec:
+            if spec.satisfies("wannier90"):
```
Comment (wdconinc) [inline var/spack/repos/builtin/packages/cp2k/package.py]:
```suggestion
            if spec.satisfies("^wannier90"):
```

File: var/spack/repos/builtin/packages/converge/package.py [hunk 9]
```diff
@@ -146,29 +146,29 @@ def install(self, spec, prefix):
...
-            if "~mpi" in spec:
+            if spec.satisfies("~mpi"):
                 converge = glob.glob("converge-*-serial*")
                 post_convert = glob.glob("post_convert_serial*")
-            elif "hp-mpi" in spec:
+            elif spec.satisfies("hp-mpi"):
                 converge = glob.glob("converge-*-hpmpi*")
                 # No HP-MPI version of post_convert
                 post_convert = glob.glob("post_convert_serial*")
             elif "intel-mpi" in spec or "intel-parallel-studio+mpi" in spec:
```
Comment (wdconinc) [inline var/spack/repos/builtin/packages/converge/package.py]: Note: this is an example of a case that is not caught.

**Review by wdconinc — COMMENTED** [review]
I started going through this (got to cpio) and found some cases where this does not apply to variants or versions, but to dependencies.

I don't think `spec.satisfies("zlib-api")` is correct (anymore?) and should be `spec.satisfies("^zlib-api")`.

There are also some cases that your script must miss due to multiple `spec in`s in a single `if`.

**Comment by aurianer** [comment]
Thank you for spotting those! I run another script to fix them and caught a few others :)

**Review by mtaillefumier — APPROVED** [review]

File: var/spack/repos/builtin/packages/converge/package.py (hunk 9 above)
Comment (wdconinc) [reply var/spack/repos/builtin/packages/converge/package.py]:
```suggestion
            elif spec.satisfies("^intel-mpi") or spec.satisfies("^intel-parallel-studio+mpi"):
```

File: var/spack/repos/builtin/packages/cosma/package.py [hunk 10]
```diff
@@ -110,11 +110,11 @@ def cosma_blas_cmake_arg(self):
     def cosma_scalapack_cmake_arg(self):
         spec = self.spec
 
-        if "~scalapack" in spec:
+        if spec.satisfies("~scalapack"):
             return "OFF"
         elif "^intel-mkl" in spec or "^intel-oneapi-mkl" in spec:
```
Comment (wdconinc) [inline var/spack/repos/builtin/packages/cosma/package.py]:
```suggestion
        elif spec.satisfied("^intel-mkl") or spec.satisfies("^intel-oneapi-mkl"):
```

File: var/spack/repos/builtin/packages/costa/package.py [hunk 11]
```diff
@@ -55,11 +55,11 @@ def setup_build_environment(self, env):
     def costa_scalapack_cmake_arg(self):
         spec = self.spec
 
-        if "~scalapack" in spec:
+        if spec.satisfies("~scalapack"):
             return "OFF"
         elif "^intel-mkl" in spec or "^intel-oneapi-mkl" in spec:
```
Comment (wdconinc) [inline var/spack/repos/builtin/packages/costa/package.py]:
```suggestion
        elif spec.satisfies("^intel-mkl") or spec.satisfies("^intel-oneapi-mkl"):
```

File: var/spack/repos/builtin/packages/cp2k/package.py [hunk 12]
```diff
@@ -426,23 +426,23 @@ def edit(self, pkg, spec, prefix):
...
+        elif spec.satisfies("%gcc"):
             fcflags += [
                 "-ffree-form",
                 "-ffree-line-length-none",
                 "-ggdb",  # make sure we get proper Fortran backtraces
             ]
-        elif "%aocc" in spec:
+        elif spec.satisfies("%aocc"):
             fcflags += ["-ffree-form", "-Mbackslash"]
         elif "%pgi" in spec or "%nvhpc" in spec:
```
Comment (wdconinc) [inline var/spack/repos/builtin/packages/cp2k/package.py]:
```suggestion
        elif spec.satisfies("%pgi") or spec.satisfies("%nvhpc"):
```

File: var/spack/repos/builtin/packages/cpio/package.py [hunk 13]
```diff
@@ -40,7 +40,7 @@ def flag_handler(self, name, flags):
         spec = self.spec
 
         if name == "cflags":
-            if "%intel@:17" in spec:
+            if spec.satisfies("%intel@:17"):
                 flags.append("-no-gcc")
 
             elif "%clang" in spec or "%fj" in spec:
```
Comment (wdconinc) [inline var/spack/repos/builtin/packages/cpio/package.py]:
```suggestion
            elif spec.satisfies("%clang") or spec.satisfies("%fj"):
```

File: var/spack/repos/builtin/packages/dihydrogen/package.py [hunk 14]
```diff
@@ -12,7 +12,7 @@
 def get_blas_entries(inspec):
     entries = []
     spec = inspec["hydrogen"]
-    if "blas=openblas" in spec:
+    if spec.satisfies("blas=openblas"):
         entries.append(cmake_cache_option("DiHydrogen_USE_OpenBLAS", True))
     elif "blas=mkl" in spec or spec.satisfies("^intel-mkl"):
```
Comment (wdconinc) [inline var/spack/repos/builtin/packages/dihydrogen/package.py]:
```suggestion
    elif spec.satisfies("blas=mkl") or spec.satisfies("^intel-mkl"):
```

File: var/spack/repos/builtin/packages/essl/package.py [hunk 15]
```diff
@@ -45,18 +45,18 @@ def blas_libs(self):
...
 
-        if "+ilp64" in spec:
+        if spec.satisfies("+ilp64"):
             essl_lib = ["libessl6464"]
         else:
             essl_lib = ["libessl"]
 
         if spec.satisfies("threads=openmp"):
             # ESSL SMP support requires XL or Clang OpenMP library
             if "%xl" in spec or "%xl_r" in spec or "%clang" in spec:
```
Comment (wdconinc) [inline var/spack/repos/builtin/packages/essl/package.py]:
```suggestion
            if spec.satisfies("%xl") or spec.satisfies("%xl_r") or spec.satisfies("%clang"):
```

**Review by wdconinc — COMMENTED** [review]
A few more missed ones (looked up to essl). They may be reflective of other patterns you can search for.

File: var/spack/repos/builtin/packages/cosma/package.py (hunk 10 above)
Comment (aurianer) [reply var/spack/repos/builtin/packages/cosma/package.py]: I rebased and pushed the last commit fixing those, my script was buggy.

**Comment by wdconinc** [comment]
@aurianer I am wondering if your scripts allow us to split this PR into a few different ones? It is hard to review (it doesn't even load properly in my browser without making the whole thing slow down to a crawl), and the slow reviews then causes the conflicts because other PRs keep getting merged. How about splitting into a few roughly equal sections alphabetically, or in different PRs by type of change (variants only, compilers only, etc). We're now at 3,263 lines. A target of < 500 per PR would be great :-) Sorry for not suggesting this earlier.

**Comment by aurianer** [comment]
I'm working on this on my free time and I admit splitting it would require me to have more time, so maybe at a later point :) I don't mind rebasing and resolving conflicts though, I already did it for this PR, it's not so much work for me.
Knowing the little change that this PR introduce, another way could be to merge it without a thorough review. If a package breaks it can be fixed instantly. Up to you to decide :)

**Comment by aurianer** [comment]
@wdconinc I've updated this PR to modify only the packages beginning with an a, I'll open PR by PR alphabetically to make them smaller, let me know if that works for you :)

**Review by wdconinc — APPROVED** [review]
Thanks. The smaller chunk is easier to review. This time I didn't explicitly look for missed cases. I did visually verify the changes instances, though.

**Pull request merged**